pkg/cert-watcher/watcher_test.go

```diff
@@ -0,0 +1,107 @@
+package certwatcher
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestCertWatcher_startWatcher(t *testing.T) {
+	type fields struct {
+		NewTLSCert string
+		NewTLSKey  string
+	}
+	tests := []struct {
+		fields fields
+	}{
+		{
+			fields: fields{
+				NewTLSCert: "dGVzdDEK",
+				NewTLSKey:  "dGVzdDEK",
+			},
+		},
+		{
+			fields: fields{
+				NewTLSCert: "dGVzdDIK",
+				NewTLSKey:  "dGVzdDIK",
+			},
+		},
+		{
+			fields: fields{
+				NewTLSCert: "dGVzdDMK",
+				NewTLSKey:  "dGVzdDMK",
+			},
+		},
+	}
+
+	tempdir := t.TempDir()
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cw := &CertWatcher{
+		TLSCertFile:     tempdir + "/tls.crt",
+		TLSKeyFile:      tempdir + "/tls.key",
+		SecretName:      "k8tz-tls",
+		SecretNamespace: "k8tz",
+		clientset:       fake.NewSimpleClientset(),
+		ctx:             ctx,
+		cancel:          cancel,
+	}
+
+	t.Run("create mock for secret resource", func(t *testing.T) {
+		if _, err := cw.clientset.CoreV1().Secrets(cw.SecretNamespace).Create(cw.ctx, &corev1.Secret{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      cw.SecretName,
+				Namespace: cw.SecretNamespace,
+			},
+			Data: map[string][]byte{
+				"tls.crt": []byte(""),
+				"tls.key": []byte(""),
+			},
+		}, v1.CreateOptions{}); err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	go func() {
+		t.Run("simulate secret resource changes", func(t *testing.T) {
+			for _, test := range tests {
+				// updating k8tz-tls secret
+				secret, err := cw.clientset.CoreV1().Secrets(cw.SecretNamespace).Get(cw.ctx, cw.SecretName, v1.GetOptions{})
+				if err != nil {
+					t.Errorf("error get secret: %v", err)
+				}
+				secret.Data["tls.crt"] = []byte(test.fields.NewTLSCert)
+				secret.Data["tls.key"] = []byte(test.fields.NewTLSKey)
+				_, err = cw.clientset.CoreV1().Secrets(cw.SecretNamespace).Update(cw.ctx, secret, v1.UpdateOptions{})
+				if err != nil {
+					t.Errorf("error update secret: %v", err)
+				}
+				time.Sleep(1 * time.Second)
+				// checking k8tz-tls secret
+				if tlsCrt, err := os.ReadFile(cw.TLSCertFile); err != nil {
+					t.Errorf("error read tls.crt: %v", err)
+				} else if string(tlsCrt) != test.fields.NewTLSCert {
+					t.Errorf("tls.crt data missmatch with data from %s: %v", cw.TLSCertFile, err)
+				}
+				if tlsKey, err := os.ReadFile(cw.TLSKeyFile); err != nil {
+					t.Errorf("error read tls.key: %v", err)
+				} else if string(tlsKey) != test.fields.NewTLSKey {
+					t.Errorf("tls.key data missmatch with data from %s: %v", cw.TLSCertFile, err)
+				}
+				time.Sleep(1 * time.Second)
```

Comment:
i am not sure that this is required

Reply:
oh ya, just realized that this is not required